src/backend/distributed/executor/local_executor.c

```diff
@@ -142,32 +143,62 @@ ExecuteLocalTaskList(CitusScanState *scanState, List *taskList)
 	foreach(taskCell, taskList)
 	{
 		Task *task = (Task *) lfirst(taskCell);
+		char *shardQueryString = NULL;
 
-		const char *shardQueryString = TaskQueryString(task);
-		Query *shardQuery = ParseQueryString(shardQueryString, parameterTypes, numParams);
+		PlannedStmt *localPlan = GetCachedLocalPlan(task, distributedPlan);
 
 		/*
-		 * We should not consider using CURSOR_OPT_FORCE_DISTRIBUTED in case of
-		 * intermediate results in the query. That'd trigger ExecuteLocalTaskPlan()
-		 * go through the distributed executor, which we do not want since the
-		 * query is already known to be local.
+		 * If the plan is already cached, don't need to re-plan, just
+		 * acquire necessary locks.
 		 */
-		int cursorOptions = 0;
+		if (localPlan != NULL)
+		{
+			shardQueryString = task->queryStringLazy
+							   ? task->queryStringLazy
+							   : "<optimized out by local execution>";
+
+			Query *jobQuery = distributedPlan->workerJob->jobQuery;
+			LOCKMODE lockMode =
+				IsModifyCommand(jobQuery) ? RowExclusiveLock : (jobQuery->hasForUpdate ?
+																RowShareLock :
+																AccessShareLock);
+
+			ListCell *oidCell = NULL;
+			foreach(oidCell, localPlan->relationOids)
+			{
+				LockRelationOid(lfirst_oid(oidCell), lockMode);
+			}
+		}
+		else
+		{
+			shardQueryString = TaskQueryString(task);
 
-		/*
-		 * Altough the shardQuery is local to this node, we prefer planner()
-		 * over standard_planner(). The primary reason for that is Citus itself
-		 * is not very tolarent standard_planner() calls that doesn't go through
-		 * distributed_planner() because of the way that restriction hooks are
-		 * implemented. So, let planner to call distributed_planner() which
-		 * eventually calls standard_planner().
-		 */
-		PlannedStmt *localPlan = planner(shardQuery, cursorOptions, paramListInfo);
+			Query *shardQuery = ParseQueryString(shardQueryString, parameterTypes,
+												 numParams);
```

Comment:
Can we use the queryForLocalExecution here?

Reply:
That would be this PR: https://github.com/citusdata/citus/pull/3368

We punted that for this release since it was not needed for the HammerDB benchmark.

Reply:
Could still be nice to include that here.

Reply:
I've tested this branch very heavily, so I'd prefer to have it as a next step, potentially in the next release. The PR #3368 will be very small on top of what we have implemented.